// aws-cdk/lib/api/util/cloudformation/stack-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StackStatus = void 0;
/**
 * A utility class to inspect CloudFormation stack statuses.
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html
 */
class StackStatus {
    constructor(name, reason) {
        this.name = name;
        this.reason = reason;
    }
    static fromStackDescription(description) {
        return new StackStatus(description.StackStatus, description.StackStatusReason);
    }
    get isCreationFailure() {
        return this.name === 'ROLLBACK_COMPLETE'
            || this.name === 'ROLLBACK_FAILED';
    }
    get isDeleted() {
        return this.name.startsWith('DELETE_');
    }
    get isFailure() {
        return this.name.endsWith('FAILED');
    }
    get isRollback() {
        return this.name.indexOf('ROLLBACK') !== -1;
    }
    get isStable() {
        return !this.name.endsWith('_IN_PROGRESS');
    }
    get isNotFound() {
        return this.name === 'NOT_FOUND';
    }
    get isSuccess() {
        return !this.isNotFound && !this.isRollback && !this.isFailure;
    }
    toString() {
        return this.name + (this.reason ? ` (${this.reason})` : '');
    }
}
exports.StackStatus = StackStatus;
