// amazon-eks-irsa-cfn/lib/identityprovider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const aws_cloudformation_1 = require("@aws-cdk/aws-cloudformation");
const aws_iam_1 = require("@aws-cdk/aws-iam");
const aws_lambda_1 = require("@aws-cdk/aws-lambda");
const path = require("path");
class OIDCIdentityProvider extends core_1.Resource {
    constructor(scope, id, props) {
        super(scope, id);
        if (!OIDCIdentityProvider.fn) {
            OIDCIdentityProvider.fn = new aws_lambda_1.Function(scope, 'OIDCIdentityProviderCustomResource', {
                code: aws_lambda_1.Code.fromAsset(path.resolve(__dirname, '..', 'lambda-packages', 'identity_provider_handler')),
                handler: 'index.handler',
                runtime: aws_lambda_1.Runtime.NODEJS_12_X,
                timeout: core_1.Duration.minutes(15),
            });
            OIDCIdentityProvider.fn.addToRolePolicy(new aws_iam_1.PolicyStatement({
                actions: [
                    'eks:DescribeCluster',
                    'iam:CreateOpenIDConnectProvider',
                    'iam:DeleteOpenIDConnectProvider'
                ],
                resources: ['*']
            }));
        }
        const provider = new aws_cloudformation_1.CustomResource(this, 'Resource', {
            provider: aws_cloudformation_1.CustomResourceProvider.fromLambda(OIDCIdentityProvider.fn),
            resourceType: 'Custom::EksOidcIdentityProvider',
            properties: {
                ClusterName: props.clusterName,
            }
        });
        this.providerArn = core_1.Token.asString(provider.getAtt('Arn'));
    }
}
exports.OIDCIdentityProvider = OIDCIdentityProvider;
