// @aws-cdk/aws-events-targets/test/batch/batch.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
const batch = require("@aws-cdk/aws-batch");
const aws_ecs_1 = require("@aws-cdk/aws-ecs");
const events = require("@aws-cdk/aws-events");
const core_1 = require("@aws-cdk/core");
const targets = require("../../lib");
test('use aws batch job as an eventrule target', () => {
    // GIVEN
    const stack = new core_1.Stack();
    const jobQueue = new batch.JobQueue(stack, 'MyQueue', {
        computeEnvironments: [
            {
                computeEnvironment: new batch.ComputeEnvironment(stack, 'ComputeEnvironment', {
                    managed: false,
                }),
                order: 1,
            },
        ],
    });
    const jobDefinition = new batch.JobDefinition(stack, 'MyJob', {
        container: {
            image: aws_ecs_1.ContainerImage.fromRegistry('test-repo'),
        },
    });
    const rule = new events.Rule(stack, 'Rule', {
        schedule: events.Schedule.expression('rate(1 min)'),
    });
    // WHEN
    rule.addTarget(new targets.BatchJob(jobQueue, jobDefinition));
    // THEN
    assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
        ScheduleExpression: 'rate(1 min)',
        State: 'ENABLED',
        Targets: [
            {
                Arn: {
                    Ref: 'MyQueueE6CA6235',
                },
                Id: 'Target0',
                RoleArn: {
                    'Fn::GetAtt': [
                        'MyJobEventsRoleCF43C336',
                        'Arn',
                    ],
                },
                BatchParameters: {
                    JobDefinition: { Ref: 'MyJob8719E923' },
                    JobName: 'Rule',
                },
            },
        ],
    }));
    assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: [
                {
                    Action: 'batch:SubmitJob',
                    Effect: 'Allow',
                    Resource: [
                        { Ref: 'MyJob8719E923' },
                        { Ref: 'MyQueueE6CA6235' },
                    ],
                },
            ],
            Version: '2012-10-17',
        },
        Roles: [
            { Ref: 'MyJobEventsRoleCF43C336' },
        ],
    }));
});
