// @aws-cdk/core/test/test.cfn-parse.js
"use strict";
const cfn_parse_1 = require("../lib/cfn-parse");
function parseCreationPolicy(policy) {
    return testCfnParser.parseCreationPolicy(policy);
}
function parseUpdatePolicy(policy) {
    return testCfnParser.parseUpdatePolicy(policy);
}
const testCfnParser = new cfn_parse_1.CfnParser({
    finder: {
        findCondition() { return undefined; },
        findRefTarget() { return undefined; },
        findResource() { return undefined; },
    },
});
module.exports = {
    'FromCloudFormation class': {
        '#parseCreationPolicy': {
            'returns undefined when given a non-object as the argument'(test) {
                test.equal(parseCreationPolicy('blah'), undefined);
                test.done();
            },
            'returns undefined when given an empty object as the argument'(test) {
                test.equal(parseCreationPolicy({}), undefined);
                test.done();
            },
            'returns undefined when given empty sub-objects as the argument'(test) {
                test.equal(parseCreationPolicy({
                    AutoScalingCreationPolicy: null,
                    ResourceSignal: {
                        Count: undefined,
                    },
                }), undefined);
                test.done();
            },
        },
        '#parseUpdatePolicy': {
            'returns undefined when given a non-object as the argument'(test) {
                test.equal(parseUpdatePolicy('blah'), undefined);
                test.done();
            },
            'returns undefined when given an empty object as the argument'(test) {
                test.equal(parseUpdatePolicy({}), undefined);
                test.done();
            },
            'returns undefined when given empty sub-objects as the argument'(test) {
                test.equal(parseUpdatePolicy({
                    AutoScalingReplacingUpdate: null,
                    AutoScalingRollingUpdate: {
                        PauseTime: undefined,
                    },
                }), undefined);
                test.done();
            },
        },
    },
};
