// @aws-cdk/aws-ecs/test/test.gelf-log-driver.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const ecs = require("../lib");
let stack;
let td;
const image = ecs.ContainerImage.fromRegistry('test-image');
module.exports = {
    'setUp'(cb) {
        stack = new cdk.Stack();
        td = new ecs.Ec2TaskDefinition(stack, 'TaskDefinition');
        cb();
    },
    'create a gelf log driver with minimum options'(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: new ecs.GelfLogDriver({
                address: 'my-gelf-address',
            }),
            memoryLimitMiB: 128,
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'gelf',
                        Options: {
                            'gelf-address': 'my-gelf-address',
                        },
                    },
                },
            ],
        }));
        test.done();
    },
    'create a gelf log driver using gelf with minimum options'(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: ecs.LogDrivers.gelf({
                address: 'my-gelf-address',
            }),
            memoryLimitMiB: 128,
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'gelf',
                        Options: {
                            'gelf-address': 'my-gelf-address',
                        },
                    },
                },
            ],
        }));
        test.done();
    },
};
