// @aws-cdk/aws-ecs/test/test.json-file-log-driver.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const ecs = require("../lib");
let stack;
let td;
const image = ecs.ContainerImage.fromRegistry('test-image');
module.exports = {
    'setUp'(cb) {
        stack = new cdk.Stack();
        td = new ecs.Ec2TaskDefinition(stack, 'TaskDefinition');
        cb();
    },
    'create a json-file log driver with options'(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: new ecs.JsonFileLogDriver({
                env: ['hello'],
            }),
            memoryLimitMiB: 128,
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'json-file',
                        Options: {
                            env: 'hello',
                        },
                    },
                },
            ],
        }));
        test.done();
    },
    'create a json-file log driver without options'(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: new ecs.JsonFileLogDriver(),
            memoryLimitMiB: 128,
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'json-file',
                    },
                },
            ],
        }));
        test.done();
    },
    'create a json-file log driver using json-file'(test) {
        // WHEN
        td.addContainer('Container', {
            image,
            logging: ecs.LogDrivers.jsonFile(),
            memoryLimitMiB: 128,
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::ECS::TaskDefinition', {
            ContainerDefinitions: [
                {
                    LogConfiguration: {
                        LogDriver: 'json-file',
                        Options: {},
                    },
                },
            ],
        }));
        test.done();
    },
};
