// @aws-cdk/aws-stepfunctions/test/activity.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
require("@aws-cdk/assert/jest");
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/core");
const stepfunctions = require("../lib");
describe('Activity', () => {
    test('instantiate Activity', () => {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new stepfunctions.Activity(stack, 'Activity');
        // THEN
        expect(stack).toHaveResource('AWS::StepFunctions::Activity', {
            Name: 'Activity',
        });
    });
    test('Activity exposes metrics', () => {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const activity = new stepfunctions.Activity(stack, 'Activity');
        // THEN
        const sharedMetric = {
            period: cdk.Duration.minutes(5),
            namespace: 'AWS/States',
            dimensions: { ActivityArn: { Ref: 'Activity04690B0A' } },
        };
        expect((stack.resolve(activity.metricRunTime()))).toEqual({
            ...sharedMetric,
            metricName: 'ActivityRunTime',
            statistic: 'Average',
        });
        expect(stack.resolve(activity.metricFailed())).toEqual({
            ...sharedMetric,
            metricName: 'ActivitiesFailed',
            statistic: 'Sum',
        });
    });
    test('Activity can grant permissions to a role', () => {
        // GIVEN
        const stack = new cdk.Stack();
        const role = new iam.Role(stack, 'Role', {
            assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        });
        const activity = new stepfunctions.Activity(stack, 'Activity');
        // WHEN
        activity.grant(role, 'states:SendTaskSuccess');
        // THEN
        expect(stack).toHaveResourceLike('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: assert_1.arrayWith(assert_1.objectLike({
                    Action: 'states:SendTaskSuccess',
                    Effect: 'Allow',
                    Resource: {
                        Ref: 'Activity04690B0A',
                    },
                })),
            },
        });
    });
});
