// @aws-cdk/aws-stepfunctions/test/custom-state.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const cdk = require("@aws-cdk/core");
const sfn = require("../lib");
const render_util_1 = require("./private/render-util");
describe('Custom State', () => {
    let stack;
    let stateJson;
    beforeEach(() => {
        // GIVEN
        stack = new cdk.Stack();
        stateJson = {
            Type: 'Task',
            Resource: 'arn:aws:states:::dynamodb:putItem',
            Parameters: {
                TableName: 'MyTable',
                Item: {
                    id: {
                        S: 'MyEntry',
                    },
                },
            },
            ResultPath: null,
        };
    });
    test('maintains the state Json provided during construction', () => {
        // WHEN
        const customState = new sfn.CustomState(stack, 'Custom', {
            stateJson,
        });
        // THEN
        expect(customState.toStateJson()).toStrictEqual({
            ...stateJson,
            End: true,
        });
    });
    test('can add a next state to the chain', () => {
        // WHEN
        const definition = new sfn.CustomState(stack, 'Custom', {
            stateJson,
        }).next(new sfn.Pass(stack, 'MyPass'));
        // THEN
        expect(render_util_1.render(stack, definition)).toStrictEqual({
            StartAt: 'Custom',
            States: {
                Custom: {
                    Next: 'MyPass',
                    Type: 'Task',
                    Resource: 'arn:aws:states:::dynamodb:putItem',
                    Parameters: {
                        TableName: 'MyTable',
                        Item: {
                            id: {
                                S: 'MyEntry',
                            },
                        },
                    },
                    ResultPath: null,
                },
                MyPass: {
                    Type: 'Pass',
                    End: true,
                },
            },
        });
    });
});
